bundle/jsky.app.ot/src/main/scala/jsky/app/ot/gemini/editor/targetComponent/details/ValidAtEditor.scala

```diff
@@ -0,0 +1,230 @@
+        delegate.getListCellRendererComponent(list, value, index, isSelected, cellHasFocus) match {
+          case label: JLabel => label <| (_.setText(value.displayValue))
+        }
+    })
+
+    addItemListener(new ItemListener {
+      override def itemStateChanged(e: ItemEvent): Unit =
+        if (e.getStateChange == ItemEvent.SELECTED) {
+          nonreentrant {
+            e.getItem match {
+              case tc: TimeConfig => setTime(tc.getDate)
+              case ts: String     => textDoc.setTime(ts) // I don't think this ever happens
+            }
+          }
+        }
+    })
+
+    textDoc.addDocumentListener(new AbstractDocumentListener {
+      override def textChanged(docEvent: DocumentEvent, newText: String): Unit =
+        nonreentrant(textDoc.setTime(newText))
+
+    })
+
+    def setTime(d: Date): Unit =
+      textDoc.setTime(timeFormatter.format(d))
+
+  }
+
+  val go   = new JButton("Go")   <| { _.addActionListener(LookupListener(plot = false)) }
+  val plot = new JButton("Plot") <| { _.addActionListener(LookupListener(plot = true))  }
+
+  /**
+   * An action listener that performs a catalog lookup, replacing the current target on success, and
+   * plotting the ephemeris if desired.
+   */
+  case class LookupListener(plot: Boolean) extends ActionListener {
+    override def actionPerformed(e: ActionEvent): Unit =
+      lookupAndSet(plot, useCache = true).runAsyncAndReportErrors
+  }
+
+  setLayout(new GridBagLayout)
+
+  add(calendar, new GridBagConstraints <| { c =>
+    c.gridx = 0
+    c.fill = GridBagConstraints.HORIZONTAL
+    c.weightx = 2
+    c.insets = new Insets(2, 2, 0, 0)
+  })
+
+  add(new JLabel("at"), new GridBagConstraints <| { c =>
+    c.gridx = 1
+    c.insets = new Insets(0, 5, 0, 0)
+  })
+
+  add(timeConfig, new GridBagConstraints <| { c =>
+    c.gridx = 2
+    c.insets = new Insets(2, 5, 0, 0)
+  })
+
+  add(new JLabel("UTC"), new GridBagConstraints <| { c =>
+    c.gridx = 3
+    c.insets = new Insets(2, 5, 0, 0)
+  })
+
+  add(go, new GridBagConstraints <| { c =>
+    c.gridx = 4
+    c.insets = new Insets(2, 5, 0, 0)
+  })
+
+  add(plot, new GridBagConstraints <| { c =>
+    c.gridx = 5
+    c.insets = new Insets(2, 2, 0, 0)
+  })
+
+  ///
+  /// METHODS
+  ///
+
+  override def edit(context: GOption[ObsContext], target: SPTarget, node0: ISPNode): Unit = {
+    spt = target
+    node = node0
+    nonreentrant {
+      val nst = spt.getTarget.asInstanceOf[NonSiderealTarget]
+      Option(nst.getDateForPosition).foreach { d =>
+        calendar.setDate(d)
+        timeConfig.setTime(d)
+      }
+    }
+  }
+
+  def objectTypeForTag(tag: ITarget.Tag): ObjectType =
+    tag match {
+      case ITarget.Tag.JPL_MINOR_BODY   => ObjectType.COMET
+      case ITarget.Tag.MPC_MINOR_PLANET => ObjectType.MINOR_BODY
+    }
+
+
+  ///
+  /// HORIZONS
+  ///
+
+  /** A program that returns the editor's current date and time. */
+  def dateTime: HorizonsIO[Date] = {
```

Comment:
Most uses of `Date` are to satisfy old APIs I think ... normally we just pass POSIX dates as `Long` values, which I find to be easier in general. But yeah that JSR is probably worth a look.